risc0/zkvm/src/host/server/exec/executor.rs

```diff
@@ -771,3 +776,58 @@ impl<'a> ExecutorImpl<'a> {
         ))
     }
 }
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ExecutorStats {
+    execution_time: Duration,
+    segment_count: usize,
+    total_prover_cycles: u64,
+    user_instruction_cycles: u64,
+    cycles_used_ratio: f64,
+}
+
+impl ExecutorStats {
+    fn new(session: &Session, elapsed: Duration) -> Result<Self> {
+        let (total_prover_cycles, user_instruction_cycles) = session.get_cycles()?;
+        let cycles_used_ratio = user_instruction_cycles as f64 / total_prover_cycles as f64 * 100.0;
+        let segment_count = session.resolve()?.len();
+
+        Ok(Self {
+            total_prover_cycles,
+            user_instruction_cycles,
+            cycles_used_ratio,
+            segment_count,
+            execution_time: elapsed,
+        })
+    }
+
+    #[instrument(name = "executor", skip(self))]
+    fn log(&self) {
+        tracing::info!("execution time: {}", self.execution_time.human_duration());
+        tracing::info!(
+            "number of segments: {}",
+            self.segment_count.human_count_bare()
+        );
+        tracing::info!(
+            "total prover cycles: {}",
+            self.total_prover_cycles.human_count_bare()
+        );
+        tracing::info!(
+            "user instruction cycles: {}",
+            self.user_instruction_cycles.human_count_bare()
+        );
+        tracing::info!(
+            "available computation used: {}%",
+            self.cycles_used_ratio.human_count_bare()
+        );
+
+        tracing::debug!(executor_stats = ?self);
+    }
+}
+
+impl TryFrom<(&Session, Duration)> for ExecutorStats {
```

Comment:
Why are we using `TryFrom`, it doesn't seem like it can possibly fail?

Reply:
`ExecutorStats::new` could fail since it calls `session.get_cycles()`, which in turn is a fallible method. To avoid panics I chose to return a `Result` instead of panicking.
(I saw in a previous comment that `.resolve()` is an expensive call, but I don't think it'd be possible to avoid incurring in its cost since `.get_cycles` uses it internally...)